src/compiler/checker.ts

```diff
@@ -5754,6 +5758,22 @@ namespace ts {
             return getNarrowedTypeOfSymbol(getExportSymbolOfValueSymbolIfExported(symbol), node);
         }
 
+        function checkEnumIdentifier(node: Node, symbol: Symbol) {
+            // const enums are only permitted in:
+            // - 'left' in property access
+            // - 'object' in indexed access
+            // - target in rhs of import statement
+            // - in a type query
+            if (isConstEnumSymbol(symbol) &&
+                !((node.parent.kind === SyntaxKind.PropertyAccessExpression && (<PropertyAccessExpression>node.parent).expression === node) ||
+                    (node.parent.kind === SyntaxKind.ElementAccessExpression && (<ElementAccessExpression>node.parent).expression === node) ||
+                    ((node.kind === SyntaxKind.Identifier || node.kind === SyntaxKind.QualifiedName) && isInRightSideOfImportOrExportAssignment(<Identifier>node)) ||
+                    isInTypeQuery(node))
+                ) {

```

Comment:
Also, instead of negating the entire expression, just `return` and report the error outside the inner `if` statement.